### PR TITLE
Path transformation to userspace

### PR DESCRIPTION
**Pull request by Moneema14**
There is a clippingpath transformation method in ParserGraphicsState.cs file. Which is a Private method and we cant use it for other any kind of path. So, to extract path(not clipping path) from pdf files we need a path transformation method. This changes allow us to transform extracting path to userspace

### CODE REVIEW

File: itext/itext.kernel/itext/kernel/geom/Path.cs [hunk 1]
```diff
@@ -245,9 +245,49 @@ public virtual void AddSubpaths<_T0>(IList<_T0> subpaths)
...
 
+        /// <summary>
+        /// Transforms the path by the specified matrix
+        /// </summary>
+        /// <param name="ctm">the matrix for the transformation</param>
+        /// <returns>the transformed path</returns>
+        public Path TransformPath(Matrix ctm)
+        {
+            Path transformedPath = new Path();
+            foreach (Subpath subpath in subpaths)
```
Comment (Excalib88) [inline itext/itext.kernel/itext/kernel/geom/Path.cs]: Why you don't use 'var'?

File: itext/itext.kernel/itext/kernel/geom/BezierCurve.cs [hunk 2]
```diff
@@ -167,5 +167,33 @@ public class BezierCurve : IShape {
...
+                AffineTransform t = new AffineTransform(ctm.Get(Matrix.I11), ctm.Get(Matrix
+                    .I12), ctm.Get(Matrix.I21), ctm.Get(Matrix.I22), ctm.Get(Matrix.I31), ctm.Get(Matrix.I32));
+                //t = t.CreateInverse();
+                Point[] transformed = new Point[points.Length];
+                t.Transform(points, 0, transformed, 0, points.Length);
+                transformedCurve = new BezierCurve(iText.IO.Util.JavaUtil.ArraysAsList(transformed));
+
+                return transformedCurve;
+            }
+            catch (Exception e)
```
Comment (Excalib88) [inline itext/itext.kernel/itext/kernel/geom/BezierCurve.cs]: need to log exception! Why you use try-catch construction and throw exception? Its not correct!

File: itext/itext.kernel/itext/kernel/geom/IShape.cs [hunk 3]
```diff
@@ -58,5 +58,12 @@ public interface IShape {
         /// consisting of shape's base points.
         /// </returns>
         IList<Point> GetBasePoints();
+        
+        /// <summary>
+        /// Transforms the shape by the specified matrix
+        /// </summary>
+        /// <param name="ctm">the matrix for the transformation</param>
+        /// <returns>the transformed shape</returns>
+        IShape TransformToUserspace(Matrix ctm);
```
Comment (Excalib88) [inline itext/itext.kernel/itext/kernel/geom/IShape.cs]: The Open Closed Principle(second SOLID principle) You need to extend abstractions! Will add a new interface with a IShape TransformToUserspace method!

File: itext/itext.kernel/itext/kernel/geom/Line.cs [hunk 4]
```diff
@@ -78,5 +79,33 @@ public Line(Point p1, Point p2)
...
+        {
+            try
+            {
+                IShape transformedLine;
+                IList<Point> basePoints = GetBasePoints();
+                Point[] points = basePoints.ToArray(new Point[basePoints.Count]);
+
+                AffineTransform t = new AffineTransform(ctm.Get(Matrix.I11), ctm.Get(Matrix
+                    .I12), ctm.Get(Matrix.I21), ctm.Get(Matrix.I22), ctm.Get(Matrix.I31), ctm.Get(Matrix.I32));
+                //t = t.CreateInverse();
```
Comment (Excalib88) [inline itext/itext.kernel/itext/kernel/geom/Line.cs]: remove commented code

File: itext/itext.kernel/itext/kernel/geom/Line.cs [hunk 5]
```diff
@@ -78,5 +79,33 @@ public Line(Point p1, Point p2)
...
+                Point[] transformed = new Point[points.Length];
+                t.Transform(points, 0, transformed, 0, points.Length);
+                transformedLine = new Line(transformed[0], transformed[1]);
+
+                return transformedLine;
+            }
+            catch (Exception e)
+            {
+                throw new Exception(e.Message, e);
```
Comment (Excalib88) [inline itext/itext.kernel/itext/kernel/geom/Line.cs]: add logging

**Comment by avlemos** [comment]
Hi @Excalib88 ,
Thank you for your contribution. Just wanted to let you know that this will get picked up by our development team (it is currently on our backlog).

**Comment by introfog** [comment]
Hi @Moneema14 
I'm from the iText Dev team
We looked at your pull request, we liked the idea, thanks for your contribution.
There are 2 comments:
1. Adding a public method to the `IShape` interface is breaking changes.
2. Duplication of code from the `ParserGraphicsState` class is not good.

As a result, we offer you the following corrections:
1. Create an `ShapeTransformUtil` class in the `com.itextpdf.kernel.geom` package, and add 3 methods (`pathTransform`, `lineTransform` and `bezierCurveTransform`).
2. Remove the methods (`transformSubpath`, `transformSegment`, `transformPoints`) in the `ParserGraphicsState` class and replace them with a method call from the `ShapeTransformUtil` class.

Also, for any non-trivial (more than 20 significant lines of code) pull request, we need a signed CLA to be in the clear regarding IP.

We are waiting for your reply.

**Comment by Snipx** [comment]
Hi @Moneema14, thanks again for the contribution and sorry for the delay. We have added a possibility to transform shapes to user space in commit https://github.com/itext/itext7-dotnet/commit/7ef86019731f915bcf2b33f3398238e979ac52f7
It's done in a different way compared to what you proposed, but that should do the trick for your use case. 

I am now going to close this pull request. Your contribution suggestions are always welcome!

Best,
Alexey

**Comment by Snipx** [comment]
Hi @Moneema14, we are looking for a way to contact you to discuss one small topic regarding your contribution. Would you be available for that? If so, could you type me at alexey.subach[at]itextpdf.com ? Thanks!

**Pull request closed without merging**